fuse/nodefs/fsops.go

```diff
@@ -70,6 +70,8 @@ func (c *FileSystemConnector) lookupMountUpdate(out *fuse.Attr, mount *fileSyste
 
 // internalLookup - execute a lookup without affecting NodeId reference counts
 func (c *FileSystemConnector) internalLookup(out *fuse.Attr, parent *Inode, name string, header *fuse.InHeader) (node *Inode, code fuse.Status) {
+	c.forgetMu.RLock()
```

Comment:
In the original code, OnForget is only called after it has successfully removed from the tree with parent.rmChild(). So, the kernel can never get a new reference to the forgotten child. Of course, that doesn't stop the pathfs from getting its parent mappings all screwed up. Maybe you can have an OnAdd() and OnRemove() which run inside the nodefs treeLock?